app/assets/javascripts/admin/addon/components/emoji-value-list.js

```diff
@@ -0,0 +1,186 @@
+  emojiPickerIsActive: false,
+  isEditorFocused: false,
+  emojiName: null,
+
+  init() {
+    this._super(...arguments);
+    this.set("collection", []);
+  },
+
+  @action
+  emojiSelected(code) {
+    this.set("emojiName", code);
+    this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    this.set("isEditorFocused", !this.isEditorFocused);
+  },
+
+  @action
+  openEmojiPicker() {
+    this.set("isEditorFocused", !this.isEditorFocused);
+    later(() => {
+      this.set("emojiPickerIsActive", !this.emojiPickerIsActive);
+    }, 100);
+  },
+
+  @action
+  clearInput() {
+    this.set("emojiName", "");
+  },
+
+  @on("didReceiveAttrs")
+  _setupCollection() {
+    this.set("collection", this._splitValues(this.values));
+  },
+
+  _splitValues(values) {
+    if (values && values.length) {
+      const res = [];
+      const emojis = values.split("|");
+      emojis.forEach((emoji, index) => {
+        const object = {};
+        object.value = emoji;
+
+        object.emojiUrl = emojiUrlFor(emoji);
+        object.isEditable = true;
+
+        object.isEditing = false;
+
+        object.isLast = emojis.length - 1 === index;
+
+        res.push(object);
+      });
+
+      return res;
+    } else {
+      return [];
+    }
+  },
+
+  @action
+  editValue(index) {
+    const item = this.collection[index];
+    if (item.isEditable) {
+      set(item, "isEditing", !item.isEditing);
+      schedule("afterRender", () => {
+        const textbox = document.querySelector(
+          `[data-index="${index}"] .value-input`
+        );
+        if (textbox) {
+          textbox.focus();
+        }
+      });
+    }
+  },
+
+  @action
+  changeValue(index, newValue) {
+    const item = this.collection[index];
+
+    if (this._checkInvalidInput(newValue)) {
+      const oldValues = this.values.split("|");
+
+      set(item, "value", oldValues[index - 1]);
+      set(item, "isEditing", !item.isEditing);
+
+      return;
+    }
+
+    this._replaceValue(index, newValue);
+
+    set(item, "isEditing", !item.isEditing);
+  },
+
+  @action
+  addValue() {
+    if (this._checkInvalidInput([this.emojiName])) {
+      return;
+    }
+    this._addValue(this.emojiName);
+    this.set("emojiName", "");
+  },
+
+  @action
+  removeValue(value) {
+    this._removeValue(value);
+  },
+
+  @action
+  shiftUp(index) {
+    if (!index) {
+      return;
+    }
+    const temp = this.collection[index];
+    this.collection[index] = this.collection[index - 1];
+    this.collection[index - 1] = temp;
+    this._saveValues();
+  },
+
+  @action
+  shiftDown(index) {
+    if (index === this.collection.length) {
+      return;
+    }
+    const temp = this.collection[index];
+    this.collection[index] = this.collection[index + 1];
+    this.collection[index + 1] = temp;
+    this._saveValues();
+  },
+
+  _checkInvalidInput(input) {
+    this.set("validationMessage", null);
+
+    if (isEmpty(input) || input.includes("|") || !emojiUrlFor(input)) {
```

Comment:
can we replace all of this by:

```
if (!emojiUrlFor(input)) {
  //
}
```

- checking input doesn’t seem necessary, `emojiUrlFor` will call `buildEmojiUrl` which does a `String(input)` so even if null or undefined will return ""
- we don’t care what input includes, we just need to know if it's a valid emoji or not

Reply:
Indeed, ill change this